go/client/getentries.go

```diff
@@ -45,19 +45,34 @@ func (c *LogClient) GetEntries(start, end int64) ([]ct.LogEntry, error) {
 	}
 	entries := make([]ct.LogEntry, len(resp.Entries))
 	for index, entry := range resp.Entries {
-		leaf, err := ct.ReadMerkleTreeLeaf(bytes.NewBuffer(entry.LeafInput))
-		if err != nil {
+		var leaf ct.MerkleTreeLeaf
+		if rest, err := tls.Unmarshal(entry.LeafInput, &leaf); err != nil {
 			return nil, err
+		} else if len(rest) > 0 {
+			return nil, fmt.Errorf("trailing data (%d bytes) after MerkleTreeLeaf", len(rest))
 		}
-		entries[index].Leaf = *leaf
+		entries[index].Leaf = leaf
 
 		var chain []ct.ASN1Cert
 		switch leaf.TimestampedEntry.EntryType {
 		case ct.X509LogEntryType:
-			chain, err = ct.UnmarshalX509ChainArray(entry.ExtraData)
+			var certChain ct.CertificateChain
+			if rest, err := tls.Unmarshal(entry.ExtraData, &certChain); err != nil {
+				return nil, err
+			} else if len(rest) > 0 {
+				return nil, fmt.Errorf("trailing data (%d bytes) after CertificateChain", len(rest))
+			}
+			chain = certChain.Entries
 
 		case ct.PrecertLogEntryType:
-			chain, err = ct.UnmarshalPrecertChainArray(entry.ExtraData)
+			var precertChain ct.PrecertChainEntry
+			if rest, err := tls.Unmarshal(entry.ExtraData, &precertChain); err != nil {
+				return nil, err
+			} else if len(rest) > 0 {
+				return nil, fmt.Errorf("trailing data (%d bytes) after PrecertChainEntry", len(rest))
+			}
+			chain = append(chain, precertChain.PreCertificate)
+			chain = append(chain, precertChain.CertificateChain...)
 
 		default:
 			return nil, fmt.Errorf("saw unknown entry type: %v", leaf.TimestampedEntry.EntryType)
```

Comment:
For some reason github won't let me comment on the lines below this, but:

I don't think the 'if err != nil { return nil, err }' that starts on line 82 in the new code is needed anymore as you handle all the errors as they arise now.